File: backend/app/schemas/equipment.py
from typing import Any, Optional, List
from datetime import date, datetime
from decimal import Decimal
from pydantic import BaseModel, Field, model_validator

from app.schemas.money import Money


class EquipmentBase(BaseModel):
    # Blank on create means "issue the next tag for this site".
    asset_tag: str = ""
    # Required for clinical equipment only (see EquipmentCreate). A chair or a
    # newly surveyed pump often has none yet; they are stored empty, not invented.
    make: str = ""
    model: str = ""
    serial_number: str = ""
    modality_id: Optional[int] = None
    facility_id: int
    tier_id: Optional[int] = None
    inspection_form_id: Optional[int] = None
    default_picture_url: Optional[str] = None
    description: Optional[str] = None
    risk_priority: Optional[str] = None
    risk_name: Optional[str] = None
    location: Optional[str] = None
    inventory_date: Optional[date] = None
    acquisition_authorized_by: Optional[str] = None
    department: Optional[str] = None
    po_no: Optional[str] = None
    requester_first_name: Optional[str] = None
    requester_last_name: Optional[str] = None
    requester_phone: Optional[str] = None
    requester_fax: Optional[str] = None
    requester_mailing_address: Optional[str] = None
    requester_email: Optional[str] = None
    owning_department: Optional[str] = None
    acquisition_method: Optional[str] = None
    acquired_company_name: Optional[str] = None
    acquired_account_number: Optional[str] = None
    acquired_sales_person: Optional[str] = None
    acquired_phone: Optional[str] = None
    acquired_email: Optional[str] = None
    acquired_mailing_address: Optional[str] = None
    cost: Optional[Decimal] = None
    acquisition_date: Optional[date] = None
    capital_equipment: Optional[str] = None
    warranty_duration: Optional[str] = None
    parts_duration: Optional[str] = None
    labor_duration: Optional[str] = None
    coverage_start_date: Optional[date] = None
    coverage_type: Optional[str] = None
    part_warranty_end_date: Optional[date] = None
    labor_warranty_end_date: Optional[date] = None
    pm_scheduling: Optional[str] = None
    installation_date: Optional[date] = None
    last_pm_date: Optional[date] = None
    next_generated_pm_date: Optional[date] = None
    purchase_date: Optional[date] = None
    warranty_expiration: Optional[date] = None
    status: str = "active"

    # ── Facilities / MEP ────────────────────────────────────────────────────
    # All optional, so every existing caller keeps working unchanged. Without
    # these an MEP asset cannot be created through the API at all: the columns
    # exist on the model but nothing can set them.
    #
    # `location` above stays as it is — the legacy free-text string, still the
    # fallback display for an asset that has not been placed in the tree.
    discipline_id: Optional[int] = None
    location_id: Optional[int] = None
    parent_equipment_id: Optional[int] = None
    criticality: Optional[str] = None
    electrical_branch: Optional[str] = None
    service_vendor_id: Optional[int] = None
    # Set for items that belong to a room (chair, display); empty for plant.
    asset_type: Optional[str] = None

    # ── Depreciation ────────────────────────────────────────────────────────
    depreciation_method: Optional[str] = None
    salvage_value: Optional[Decimal] = None
    useful_life_years: Optional[Decimal] = None
    total_expected_units: Optional[Decimal] = None


class RoomAssetsCreate(BaseModel):
    """Add several of one item to a room, each as its own asset."""

    location_id: int
    asset_type: str
    count: int = Field(default=1, ge=1, le=200)
    # Only for a type the catalogue does not know: which trade maintains it.
    discipline_code: Optional[str] = None
    # Only for a single item: the tag already on it, and its serial.
    asset_tag: Optional[str] = None
    serial_number: Optional[str] = None
    # Shared by every item added together.
    make: Optional[str] = None
    model: Optional[str] = None
    cost: Optional[Decimal] = None
    installation_date: Optional[date] = None
    description: Optional[str] = None


class AssetSelection(BaseModel):
    """Which assets a bulk change applies to: ticked ones, or everything a filter matches.

    The filter mirrors the register's, so "select all 340 matching" means the
    same 340 the list said, not just the hundred loaded on screen. A filter
    must name a site: a bulk change never spans hospitals.
    """

    ids: Optional[List[int]] = Field(default=None, max_length=5000)
    facility_id: Optional[int] = None
    search: Optional[str] = None
    location_id: Optional[int] = None
    kind: Optional[str] = Field(default=None, pattern="^(room_items|equipment)$")
    asset_type: Optional[str] = None
    discipline_id: Optional[int] = None

    @model_validator(mode="after")
    def one_way_of_choosing(self):
        if self.ids is not None and self.facility_id is not None:
            raise ValueError("Choose assets either by ticking them or by filter, not both")
        if self.ids is None and self.facility_id is None:
            raise ValueError("Say which assets: tick them, or filter within a site")
        if self.ids is not None and not self.ids:
            raise ValueError("No assets are selected")
        return self


class AssetBulkChanges(BaseModel):
    """The details to set. Anything left out or blank is left as it is."""

    cost: Optional[Money] = None
    installation_date: Optional[date] = None
    make: Optional[str] = Field(default=None, max_length=200)
    model: Optional[str] = Field(default=None, max_length=200)
    discipline_id: Optional[int] = None


class AssetBulkUpdate(BaseModel):
    selection: AssetSelection
    changes: AssetBulkChanges
    # Preview by default: the caller has to ask for the change to happen.
    dry_run: bool = True


class BulkSkipped(BaseModel):
    id: int
    asset_tag: str
    reason: str


class BulkFieldOutcome(BaseModel):
    field: str
    label: str
    value: Any
    will_change: int
    unchanged: int
    skipped_count: int
    skipped: List[BulkSkipped]


class AssetBulkResult(BaseModel):
    dry_run: bool
    matched: int
    assets_changed: int
    fields: List[BulkFieldOutcome]


class ServesSpace(BaseModel):
    """One space an asset supplies, and with what."""

    location_id: int
    service_type: str


class ServesLink(BaseModel):
    id: int
    location_id: int
    code: str
    name: Optional[str] = None
    location_type: str
    criticality: Optional[str] = None
    service_type: str


class EquipmentCreate(EquipmentBase):
    # The spaces it supplies: an air handler in a roof plant room serving the
    # theatres. Also decides its criticality when none is given.
    serves: List[ServesSpace] = []

    @model_validator(mode="before")
    @classmethod
    def blanks_not_nulls(cls, data):
        if isinstance(data, dict):
            for key in ("asset_tag", "make", "model", "serial_number"):
                if data.get(key) is None:
                    data[key] = ""
        return data

    @model_validator(mode="after")
    def clinical_equipment_is_identified(self):
        """Make, model and serial are how a recall reaches a device."""
        if self.modality_id is not None:
            missing = [label for label, value in (
                ("make", self.make), ("model", self.model), ("serial number", self.serial_number),
            ) if not value.strip()]
            if missing:
                raise ValueError(
                    f"Clinical equipment needs its {', '.join(missing)}: recalls and "
                    "safety notices are tracked by them."
                )
        return self

    @model_validator(mode="after")
    def needs_a_classification(self):
        """One of modality or discipline, not both and not neither.

        A ventilator is classified clinically; a lift is classified by trade.
        Leaving both empty produces an asset that no maintenance programme,
        dispatch rule or depreciation default can reason about, which is worse
        than refusing it.
        """
        if self.modality_id is None and self.discipline_id is None:
            raise ValueError(
                "Give the asset either a modality (clinical equipment) or a "
                "discipline (plant and MEP)."
            )
        return self


class EquipmentUpdate(BaseModel):
    asset_tag: Optional[str] = None
    make: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    modality_id: Optional[int] = None
    facility_id: Optional[int] = None
    tier_id: Optional[int] = None
    inspection_form_id: Optional[int] = None
    default_picture_url: Optional[str] = None
    description: Optional[str] = None
    risk_priority: Optional[str] = None
    risk_name: Optional[str] = None
    location: Optional[str] = None
    inventory_date: Optional[date] = None
    acquisition_authorized_by: Optional[str] = None
    department: Optional[str] = None
    po_no: Optional[str] = None
    requester_first_name: Optional[str] = None
    requester_last_name: Optional[str] = None
    requester_phone: Optional[str] = None
    requester_fax: Optional[str] = None
    requester_mailing_address: Optional[str] = None
    requester_email: Optional[str] = None
    owning_department: Optional[str] = None
    acquisition_method: Optional[str] = None
    acquired_company_name: Optional[str] = None
    acquired_account_number: Optional[str] = None
    acquired_sales_person: Optional[str] = None
    acquired_phone: Optional[str] = None
    acquired_email: Optional[str] = None
    acquired_mailing_address: Optional[str] = None
    cost: Optional[Decimal] = None
    acquisition_date: Optional[date] = None
    capital_equipment: Optional[str] = None
    warranty_duration: Optional[str] = None
    parts_duration: Optional[str] = None
    labor_duration: Optional[str] = None
    coverage_start_date: Optional[date] = None
    coverage_type: Optional[str] = None
    part_warranty_end_date: Optional[date] = None
    labor_warranty_end_date: Optional[date] = None
    pm_scheduling: Optional[str] = None
    installation_date: Optional[date] = None
    last_pm_date: Optional[date] = None
    next_generated_pm_date: Optional[date] = None
    purchase_date: Optional[date] = None
    warranty_expiration: Optional[date] = None
    status: Optional[str] = None

    # ── Facilities / MEP ────────────────────────────────────────────────────
    # All optional, so every existing caller keeps working unchanged. Without
    # these an MEP asset cannot be created through the API at all: the columns
    # exist on the model but nothing can set them.
    #
    # `location` above stays as it is — the legacy free-text string, still the
    # fallback display for an asset that has not been placed in the tree.
    discipline_id: Optional[int] = None
    location_id: Optional[int] = None
    parent_equipment_id: Optional[int] = None
    criticality: Optional[str] = None
    electrical_branch: Optional[str] = None
    service_vendor_id: Optional[int] = None
    asset_type: Optional[str] = None

    # ── Depreciation ────────────────────────────────────────────────────────
    depreciation_method: Optional[str] = None
    salvage_value: Optional[Decimal] = None
    useful_life_years: Optional[Decimal] = None
    total_expected_units: Optional[Decimal] = None


class Equipment(EquipmentBase):
    id: int
    created_at: datetime
    updated_at: datetime
    # "Chair", "Ceiling speaker": what a room item is, for lists that would
    # otherwise show an empty make and model.
    type_label: Optional[str] = None
    # Set on equipment in the Facility Categories (see app/services/site_categories.py),
    # so the register can show it by name and where exactly it is.
    name: Optional[str] = None
    equipment_type: Optional[str] = None
    quantity: Optional[int] = None
    building: Optional[str] = None
    floor: Optional[str] = None
    condition: Optional[str] = None

    class Config:
        from_attributes = True


class EquipmentListResponse(BaseModel):
    items: List[Equipment]
    total: int

File: frontend/src/api/equipment.ts
import apiClient from './client'

export interface EquipmentItem {
  id: number
  asset_tag: string
  make: string
  model: string
  serial_number: string
  modality_id: number
  facility_id: number
  tier_id: number | null
  inspection_form_id: number | null
  default_picture_url: string | null
  description: string | null
  risk_priority: string | null
  risk_name: string | null
  location: string | null
  inventory_date: string | null
  acquisition_authorized_by: string | null
  department: string | null
  po_no: string | null
  requester_first_name: string | null
  requester_last_name: string | null
  requester_phone: string | null
  requester_fax: string | null
  requester_mailing_address: string | null
  requester_email: string | null
  owning_department: string | null
  acquisition_method: string | null
  acquired_company_name: string | null
  acquired_account_number: string | null
  acquired_sales_person: string | null
  acquired_phone: string | null
  acquired_email: string | null
  acquired_mailing_address: string | null
  cost: number | null
  acquisition_date: string | null
  capital_equipment: string | null
  warranty_duration: string | null
  parts_duration: string | null
  labor_duration: string | null
  coverage_start_date: string | null
  coverage_type: string | null
  part_warranty_end_date: string | null
  labor_warranty_end_date: string | null
  pm_scheduling: string | null
  installation_date: string | null

  // Depreciation. The backend has accepted these since the asset ledger
  // landed; they were missing here, which is why no form could offer them.
  depreciation_method: string | null
  salvage_value: string | number | null
  useful_life_years: string | number | null
  total_expected_units: string | number | null
  last_pm_date: string | null
  next_generated_pm_date: string | null
  purchase_date: string | null
  warranty_expiration: string | null
  status: string
  created_at: string
  updated_at: string

  // Facilities placement. A room item (chair, display) has an asset_type and a
  // type_label; a lift or a ventilator has neither.
  location_id?: number | null
  discipline_id?: number | null
  criticality?: string | null
  asset_type?: string | null
  type_label?: string | null

  // Set on equipment in the Facility Categories: known by name, and where
  // exactly it is. `location` holds the room or exact spot.
  name?: string | null
  equipment_type?: string | null
  quantity?: number | null
  building?: string | null
  floor?: string | null
  condition?: string | null
}

export interface EquipmentCreate {
  /** Blank: the next tag for the site is issued. */
  asset_tag?: string
  /** Required for clinical equipment only. */
  make?: string
  model?: string
  serial_number?: string
  modality_id?: number | null
  discipline_id?: number | null
  location_id?: number | null
  criticality?: string | null
  asset_type?: string | null
  serves?: Array<{ location_id: number; service_type: string }>
  facility_id: number
  tier_id?: number | null
  inspection_form_id?: number | null
  default_picture_url?: string
  description?: string
  risk_priority?: string
  risk_name?: string
  location?: string
  inventory_date?: string | null
  acquisition_authorized_by?: string
  department?: string
  po_no?: string
  requester_first_name?: string
  requester_last_name?: string
  requester_phone?: string
  requester_fax?: string
  requester_mailing_address?: string
  requester_email?: string
  owning_department?: string
  acquisition_method?: string
  acquired_company_name?: string
  acquired_account_number?: string
  acquired_sales_person?: string
  acquired_phone?: string
  acquired_email?: string
  acquired_mailing_address?: string
  cost?: number
  acquisition_date?: string | null
  capital_equipment?: string
  warranty_duration?: string
  parts_duration?: string
  labor_duration?: string
  coverage_start_date?: string | null
  coverage_type?: string
  part_warranty_end_date?: string | null
  labor_warranty_end_date?: string | null
  pm_scheduling?: string
  installation_date?: string | null

  // Depreciation. The backend has accepted these since the asset ledger
  // landed; they were missing here, which is why no form could offer them.
  depreciation_method?: string | null
  salvage_value?: string | number | null
  useful_life_years?: string | number | null
  total_expected_units?: string | number | null
  last_pm_date?: string | null
  next_generated_pm_date?: string | null
  purchase_date?: string
  warranty_expiration?: string
  status?: string
}

export interface EquipmentListResponse {
  items: EquipmentItem[]
  total: number
}

export const fetchEquipment = async (facilityId?: number, search?: string): Promise<EquipmentListResponse> => {
  const params = facilityId
    ? { facility_id: facilityId, search: search || undefined, limit: 500 }
    : { search: search || undefined, limit: 500 }
  const res = await apiClient.get('/equipment/', { params })
  return res.data
}

export interface EquipmentUpdate extends Partial<EquipmentCreate> {}

/**
 * The asset register for one site, filtered on the server.
 *
 * The Assets page used to call fetchEquipment with an object where it takes a
 * facility id, so the site filter never reached the server and every site's
 * register listed every site's assets. Named parameters here make that a type
 * error rather than a silent leak.
 */
export const fetchAssetRegister = async (params: {
  /** Required by the register page; a room's own list is scoped by location_id. */
  facility_id?: number
  search?: string
  location_id?: number | null
  kind?: 'room_items' | 'equipment' | null
  discipline_id?: number | null
  skip?: number
  limit?: number
}): Promise<{ items: EquipmentItem[]; total: number }> => {
  const res = await apiClient.get('/equipment/', {
    params: {
      facility_id: params.facility_id,
      search: params.search || undefined,
      location_id: params.location_id ?? undefined,
      kind: params.kind ?? undefined,
      discipline_id: params.discipline_id ?? undefined,
      skip: params.skip ?? 0,
      limit: params.limit ?? 100,
    },
  })
  return res.data
}

export const fetchEquipmentById = async (id: number): Promise<EquipmentItem> => {
  const res = await apiClient.get(`/equipment/${id}`)
  return res.data
}

export interface RoomItemType {
  key: string
  label: string
  discipline: string
  discipline_id: number | null
}

export const fetchRoomItemTypes = async (): Promise<{ types: RoomItemType[] }> => {
  const res = await apiClient.get('/equipment/room-item-types')
  return res.data
}

/** Put several of one item in a room, each its own asset with its own tag. */
export const addRoomItems = async (payload: {
  location_id: number
  asset_type: string
  count: number
  discipline_code?: string | null
  /** Only for a single item. */
  asset_tag?: string | null
  serial_number?: string | null
  make?: string | null
  model?: string | null
  cost?: number | null
  installation_date?: string | null
  description?: string | null
}): Promise<{ items: EquipmentItem[]; total: number }> => {
  const res = await apiClient.post('/equipment/room-items', payload)
  return res.data
}

/** Which assets a bulk change covers: ticked ones, or everything a register filter matches. */
export type AssetSelection =
  | { ids: number[] }
  | {
      facility_id: number
      search?: string
      location_id?: number | null
      kind?: 'room_items' | 'equipment' | null
      discipline_id?: number | null
    }

/** Details to set in bulk. Leave a field out to leave it as it is. */
export interface AssetBulkChanges {
  cost?: number
  installation_date?: string
  make?: string
  model?: string
  discipline_id?: number
}

export interface AssetBulkResult {
  dry_run: boolean
  matched: number
  assets_changed: number
  fields: Array<{
    field: keyof AssetBulkChanges
    label: string
    value: unknown
    will_change: number
    unchanged: number
    skipped_count: number
    skipped: Array<{ id: number; asset_tag: string; reason: string }>
  }>
}

/** Preview (the default) or apply the same details to many assets. */
export const bulkUpdateAssets = async (payload: {
  selection: AssetSelection
  changes: AssetBulkChanges
  dry_run: boolean
}): Promise<AssetBulkResult> => {
  const res = await apiClient.post('/equipment/bulk-update', payload)
  return res.data
}

/** The tag the next registration at this site will get, for the form to show. */
export const fetchNextTag = async (facilityId: number): Promise<string> => {
  const res = await apiClient.get('/equipment/next-tag', { params: { facility_id: facilityId } })
  return res.data.tag
}

export interface ServesLink {
  id: number
  location_id: number
  code: string
  name: string | null
  location_type: string
  criticality: string | null
  service_type: string
}

/** The spaces an asset supplies — an air handler's theatres, a panel's rooms. */
export const fetchServes = async (equipmentId: number): Promise<ServesLink[]> => {
  const res = await apiClient.get(`/equipment/${equipmentId}/serves`)
  return res.data
}

export const addServes = async (
  equipmentId: number, payload: { location_id: number; service_type: string },
): Promise<ServesLink[]> => {
  const res = await apiClient.post(`/equipment/${equipmentId}/serves`, payload)
  return res.data
}

export const removeServes = async (equipmentId: number, linkId: number): Promise<ServesLink[]> => {
  const res = await apiClient.delete(`/equipment/${equipmentId}/serves/${linkId}`)
  return res.data
}

export const createEquipment = async (data: EquipmentCreate): Promise<EquipmentItem> => {
  const res = await apiClient.post('/equipment/', data)
  return res.data
}

export const updateEquipment = async (id: number, data: EquipmentUpdate): Promise<EquipmentItem> => {
  const res = await apiClient.put(`/equipment/${id}`, data)
  return res.data
}

export const deleteEquipment = async (id: number): Promise<void> => {
  await apiClient.delete(`/equipment/${id}`)
}

export const exportEquipmentCsv = async (): Promise<void> => {
  const res = await apiClient.get('/equipment/export-csv', { responseType: 'blob' })
  const blob = new Blob([res.data], { type: 'text/csv' })
  const url = window.URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = 'facility_inventory.csv'
  document.body.appendChild(a)
  a.click()
  window.URL.revokeObjectURL(url)
  a.remove()
}
